Source/FieldSolver/SpectralSolver/SpectralAlgorithms/PMLPsatdAlgorithmRZ.cpp

```diff
@@ -0,0 +1,176 @@
+        int const nr = bx.length(0);
+
+        // Loop over indices within one box
+        // Note that k = 0
+        int const modes = f.n_rz_azimuthal_modes;
+        amrex::ParallelFor(bx, modes,
+        [=] AMREX_GPU_DEVICE(int i, int j, int k, int mode) noexcept
+        {
+
+            // All of the fields of each mode are grouped together
+            int const Ep_m_pml = Idx.Er_pml + Idx.n_fields*mode;
+            int const Em_m_pml = Idx.Et_pml + Idx.n_fields*mode;
+            int const Bp_m_pml = Idx.Br_pml + Idx.n_fields*mode;
+            int const Bm_m_pml = Idx.Bt_pml + Idx.n_fields*mode;
+            int const Ez_m = Idx.Ez + Idx.n_fields*mode;
+            int const Bz_m = Idx.Bz + Idx.n_fields*mode;
+
+            // Record old values of the fields to be updated
+            Complex const Ep_old_pml = fields(i,j,k,Ep_m_pml);
+            Complex const Em_old_pml = fields(i,j,k,Em_m_pml);
+            Complex const Bp_old_pml = fields(i,j,k,Bp_m_pml);
+            Complex const Bm_old_pml = fields(i,j,k,Bm_m_pml);
+            Complex const Ez_old = fields(i,j,k,Ez_m);
+            Complex const Bz_old = fields(i,j,k,Bz_m);
+
+            // k vector values, and coefficients
+            // The k values for each mode are grouped together
+            int const ir = i + nr*mode;
+            amrex::Real const kr = kr_arr[ir];
+
+            constexpr amrex::Real c2 = PhysConst::c*PhysConst::c;
+            Complex const I = Complex{0._rt,1._rt};
+            amrex::Real const C = C_arr(i,j,k,mode);
+            amrex::Real const S_ck = S_ck_arr(i,j,k,mode);
+
+            // Update E (see WarpX online documentation: theory section)
+            fields(i,j,k,Ep_m_pml) = C*Ep_old_pml
+                        + S_ck*(-c2*I*kr/2._rt*Bz_old);
+            fields(i,j,k,Em_m_pml) = C*Em_old_pml
+                        + S_ck*(-c2*I*kr/2._rt*Bz_old);
+            // Update B (see WarpX online documentation: theory section)
+            fields(i,j,k,Bp_m_pml) = C*Bp_old_pml
+                        - S_ck*(-I*kr/2._rt*Ez_old);
+            fields(i,j,k,Bm_m_pml) = C*Bm_old_pml
+                        - S_ck*(-I*kr/2._rt*Ez_old);
+
+        });
+    }
+}
+
+void PMLPsatdAlgorithmRZ::InitializeSpectralCoefficients (SpectralFieldDataRZ const & f)
+{
+
+    // Fill them with the right values:
+    // Loop over boxes and allocate the corresponding coefficients
+    // for each box owned by the local MPI proc
+    for (amrex::MFIter mfi(f.fields); mfi.isValid(); ++mfi){
+
+        amrex::Box const & bx = f.fields[mfi].box();
+
+        // Extract pointers for the k vectors
+        amrex::Real const* const modified_kz = modified_kz_vec[mfi].dataPtr();
+
+        // Extract arrays for the coefficients
+        amrex::Array4<amrex::Real> const& C = C_coef[mfi].array();
+        amrex::Array4<amrex::Real> const& S_ck = S_ck_coef[mfi].array();
+
+        auto const & kr_modes = f.getKrArray(mfi);
```

Comment:
This could work like this, but better to double check:
```suggestion
        RealVector const & kr_modes = f.getKrArray(mfi);
```

Reply:
Same here